Wrappers/Python/cil/processors/PaganinProcessor.py

```diff
@@ -0,0 +1,537 @@
+    UNIT_LIST = ['m','cm','mm','um']
+    UNIT_MULTIPLIERS = [1.0, 1e-2, 1e-3, 1e-6]
+   
+    def __init__(self, delta=1, beta=1e-2, energy=40000,  full_retrieval=True, 
+                 filter_type='paganin_method', pad=0, return_units='cm'):
+
+        if return_units not in self.UNIT_LIST:
+            raise ValueError("Return units {} not recognised, expected one of \
+                             {}, please update data.geometry.config.units"
+                             .format(str(return_units),str(self.UNIT_LIST)))
+        else:
+            return_multiplier = 1/(self.UNIT_MULTIPLIERS[self.UNIT_LIST.index
+                                                         (return_units)])
+        
+        kwargs = {
+            'energy' : energy,
+            'wavelength' : self._energy_to_wavelength(energy, 
+                                                      return_multiplier),
+            'delta': delta,
+            'beta': beta,
+            '_delta_user' : delta,
+            '_beta_user' : beta,
+            'filter_type' : filter_type,
+            'mu' : None,
+            'alpha' : None,
+            'pixel_size' : None,
+            'propagation_distance' : None,
+            'magnification' : None,
+            'filter' : None,
+            'full_retrieval' : full_retrieval,
+            'pad' : pad,
+            'override_geometry' : None,
+            'override_filter' : None,
+            'return_multiplier' : return_multiplier
+            }
+        
+        super(PaganinProcessor, self).__init__(**kwargs)
+
+    def check_input(self, data):
+        if not isinstance(data, (AcquisitionData)):
+            raise TypeError('Processor only supports AcquisitionData')
+      
+        return True
+        
+    def process(self, out=None):
+
+        data  = self.get_input()
+
+        # set the geometry parameters to use from data.geometry unless the 
+        # geometry is overridden with an override_geometry
+        self._set_geometry(data.geometry, self.override_geometry)
+
+        if out is None:
+            out = data.geometry.allocate(None)
+
+        if self.pad>0:
+            try:
+                # pad the data
+                Padder.edge(pad_width={'horizontal': self.pad, 
+                                       'vertical':self.pad})(data)
+                Padder.edge(pad_width={'horizontal': self.pad, 
+                                       'vertical':self.pad})(out)
+            except:
+                # pad just in horizontal if the data is 2D
+                Padder.edge(pad_width={'horizontal': self.pad})(data)
+                Padder.edge(pad_width={'horizontal': self.pad})(out)
+
+        # create an empty axis if the data is 2D
+        if data.geometry.dimension == '2D':
+            data = np.expand_dims(data.as_array(),2)
+        else:
+            data = data.as_array()
+
+        # create a filter based on the shape of the data
+        filter_shape = np.shape(data.take(indices = 0, axis = 
```

Comment:
Does this work if the data is ordered differently initially? say 'horizontal, vertical, angles' I suppose we want to require the input in a certain order initially.

Reply:
Yes this should work, I've added a test